// @webassemblyjs/validation/esm/index.js
function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

import importOrderValidate from "./import-order";
import isConst from "./is-const";
import typeChecker from "./type-checker";
import imports from "./imports";
import duplicatedExports from "./duplicated-exports";
import { moduleContextFromModuleAST } from "@webassemblyjs/ast";
export default function validateAST(ast) {
  var errors = getValidationErrors(ast);

  if (errors.length !== 0) {
    var errorMessage = "Validation errors:\n" + errors.join("\n");
    throw new Error(errorMessage);
  }
}
export function getValidationErrors(ast) {
  var errors = [];
  var modules = []; // $FlowIgnore

  if (ast.type === "Module") {
    modules = [ast];
  } // $FlowIgnore


  if (ast.type === "Program") {
    modules = ast.body.filter(function (_ref) {
      var type = _ref.type;
      return type === "Module";
    });
  }

  modules.forEach(function (m) {
    var moduleContext = moduleContextFromModuleAST(m); // $FlowIgnore

    errors.push.apply(errors, _toConsumableArray(imports(ast, moduleContext)));
    errors.push.apply(errors, _toConsumableArray(isConst(ast, moduleContext)));
    errors.push.apply(errors, _toConsumableArray(importOrderValidate(ast)));
    errors.push.apply(errors, _toConsumableArray(typeChecker(ast, moduleContext)));
    errors.push.apply(errors, _toConsumableArray(duplicatedExports(ast)));
  });
  return errors;
}
export { getType, typeEq } from "./type-inference";
export { isConst };
export var stack = typeChecker;
